### ISSUE TITLE
Invalid operators are ignored by Java Decaf instead of producing an error

### ISSUE THREAD

**Issue by sergutsan** [SophieKoonin/JavaDecaf#16]
**Summary of the problem:**

When the Java Decaf compiler encounters an invalid operator, it ignores it instead of producing an error. 

**What do you see?**

Code like 

```
String str = readLine();
if (str.length() >> 5) {
  println();
}
```

produces the following output: 

```
Hello.java:9: error: ')' expected
    if (str.length() 5) {
                    ^
Hello.java:9: error: not a statement
    if (str.length() 5) {
                     ^
Hello.java:9: error: ';' expected
    if (str.length() 5) {
                      ^
3 errors
```

**What should you see instead?**

Something like: 

```
Hello.java:2: error: '>>' is not a valid operator for this type of data. 
    if (str.length() >> 5) {
                     ^
1 error
```

**Comment by sergutsan** [comment]
See also Issue #15 about errors and line numbers.

**Comment by sophiekoonin** [comment]
Not sure why it's ignoring it - the compiler shouldn't technically be ignoring anything, just printing it verbatim and letting javac pick it up.

**Comment by sophiekoonin** [comment]
aha! operators like >> and << are considered a "shift expression". Since we can probably assume students won't be using these, maybe we can create a new AST node for these operators and throw an InvalidOperatorException when one is detected?

Current:

```
void ShiftExpression() :
{} 
{
  AdditiveExpression() (  ( "<<" | ">>" | ">>>" ) AdditiveExpression() )*
}
```

Suggested:

```
void ShiftExpression() :
{Token t;} 
{
  AdditiveExpression() (  ({
t = getToken(1);
}
(( "<<" | ">>" | ">>>" ) 
{ setFirstLastToken(jjtThis, t, getToken(1));}))#BadOperator AdditiveExpression() )*
}
```

In the body of the BadOperator node, there could just be a call to add a new error to the error list.

**Comment by sergutsan** [comment]
Thanks, Sophie. Will try that. :-)

Could you please explain why these operators are suppressed from the output instead of being left verbatim? Maybe there are similar problems lurking in other rarely-used operators...?

**Comment by sophiekoonin** [comment]
It's to do with how the conditional operators are processed - the condition
is split into Expression() Operator() Expression() (simplified for
explanation) and if no operator node is defined it won't be added to the
list of children to process. Each node only processes its own content, so
if any content such as an operator 'slips through the cracks' it will be
forgotten.

(JavaCC is horrible.)

I made a decision not to create nodes for every single structure in the
grammar since I thought some were unlikely to be used.

**Comment by sergutsan** [comment]
I see. In a context of limited resources (time) that is a sensible decision, given its low return-of-investment compared to other tasks. 

In the long run, though, I would argue that it is better not to remove anything from the students´ code to prevent confusion. IMHO JavaDecaf should either make a comment when pedagogically sensible or just dump whatever there is and let Java deal with it, but not remove it. Does this make sense?

If yes, what would be the resulting code in this case? Maybe something similar to...?

```
void ShiftExpression() :
{} 
{
    AdditiveExpression() (  
        ( 
            { t = getToken(1); }
            (
                ( "<<" | ">>" | ">>>" ) 
                { setFirstLastToken(jjtThis, t, getToken(1)); } 
            )
         )
        AdditiveExpression() 
    )*
}
```

**Comment by sophiekoonin** [comment]
Yep - see my above comment - https://github.com/SophieKoonin/JavaDecaf/issues/16#issuecomment-150147733

**Comment by sergutsan** [comment]
That´s the comment I used as a model given my limited current knowledge of JavaCC. :-)

Looking at it again, I understand that the text after # is the name of the class that JavaCC will create (extending SimpleNode). Is this correct? If it is, my comment before was wrong. 

I seem to understand now that the place where it is decided what to do with a token is in the ASTX class, right?, regardless of whether this means ignoring, copying, showing error, or any other thing... Modifying the syntax only produces ASTX classes, it does not affect the behaviour of the parser with respect to the tokens. 

In other words, if we just want to copy this shift-operator, we only need to (1) modify the syntax as per your comment above, (2) rebuild the tree, (3) re-generate the AST classes, resulting in a brand-new ASTXxxx class, and (4) nothing else because the default behaviour of a SimpleNode is to write itself. Did I miss something important?

**Comment by sophiekoonin** [comment]
Yes, that's right. so the #NodeName bit signals to JJTree/JavaCC that it
needs to create a new node class.
